### ISSUE TITLE
Doesn't seem to be working for me

### ISSUE THREAD

**Issue by flmbray** [AraComITServicesAG/insomnia-plugin-oauth2-token-provider#1]
I have two requests - request A and request B.  Request A uses OAuth2.0 and works fine - it gets the token, and can get the data and display the response.   Request B is configured to use Bearer authorization, and uses Token = `access_token`.  But when the request is sent, all I see in the Timeline is "Authorization: Bearer " (yes, I think there is a space at the end.)  The plugin version is 1.0.2, and does show as "enabled".  Insomnia Version is 2021.3.0.

**Comment by kitingChris** [comment]
Did you fetch the token in the OAuth2 from your login request in advance? 

![image](https://user-images.githubusercontent.com/757762/116971431-d5c02a00-acb9-11eb-852c-a1fa7f69779d.png)

The button "Fetch Tokens" should open your login page and refresh token and access token should be filled.

**Comment by flmbray** [comment]
Well, yes, but...   I don't know how to say this without seeming like a noob, but it turns out I just had to set the correct oauth2_token_url_suffix.  The full story of how I figured that out (even though I _had_ read the documentation, promise!) would explain how your image actually helped me quite a bit, but would also require a number of screenshots and probably would only make me sound like a rambling drunk.  Instead I'll just say that seeing that you had a separate request for the /Identity/Account/Login lead me to a better understanding (although I understand it's not actually necessary).  If you really want the whole story, I'm happy to detail it, but don't say I didn't warn you.  😉

**Comment by kitingChris** [comment]
Sure! Would be happy for any remark to enhance the Readme ;-)

**Comment by flmbray** [comment]
After a few days, and now that I know the proper procedure, I'm finding it hard to recall exactly how things went, so apologies if some of this doesn't seem to flow smoothly, and TBH, I'm writing some of this while I'm trying to recreate my steps, so it may not be completely accurate to how things actually went.  I don't think there's anything _wrong_ with the README, but when I read the bit about different auth urls, it just didn't click that it applied to my situation (explained below.)  But I think I can still provide some suggestions for improvement for the README and/or the product itself.  I'm going to try to do this without screenshots to keep the post length down.

First, know that I was actually trying to figure out the issue for a colleague  In my own work, I don't normally use OAuth in the Auth tab - I write a separate request and then use Insomnia's built-in ability to get data from a response.  But my colleague was asking about re-using the token from the OAuth tab, which is what led me to your product.  So I had her export her stuff and send it to me, which I then imported into a new workspace.  I don't recall whether I had added the plugin already before this, or if I added it after.

Since I was expecting the plugin to use the similar functionality as Insomnia's built-in linking, I was at first confused about how the plugin would get the data when it didn't let me choose a specific request from which to get the data - that is, when adding the 'access_token' tag in the Bearer token (on "request B") I was expecting to have to choose a specific request.  But the tag description clearly indicated that it would return the "last found OAuth2 token".  I already had a token from my OAuth request ("request A"), but the Live Preview was showing "error".  So then I guessed that maybe I had to get the token after the tag was configured.  So I went back to the request A's OAuth tab, cleared the token, fetched the token, and the new token displayed.  Then I returned to request B, clicked the 'access_token' tab and fully expected to see the token in the Live Preview, but it still said error.  Clicking 'refresh' switched it briefly to 'rendering...' but then right back to 'error' even though I had just gotten the token.

I think it was at this point that I submitted my my original question above.

After seeing your response with the screenshot, I thought "Ok maybe it does actually need a separate request", so I built one as a direct request to the OAuth URL, but it still didn't work.   This is where I started to feel a bit doofy, because my first thought was that it was weird that you had an "account login" URL with OAuth - this isn't needed on API's I've had experience with, which only require the OAuth token to access any API covered by the token (i.e. there is no separate "login" for those I've dealt with.)  I understand that's not necessarily true for all APIs, but for whatever reason, this thought was what triggered my memory of reading about the oauth2_token_url_suffix. 

So I added the oauth2_token_url_suffix variable to the environment, then went back to the request B access_token link, and again I expected it to work, but it was still showing "error" and refresh still didn't work.  I went and refreshed the request A token again, then back to request B access_token, and it still didn't work, and refresh still didn't work.   I was all ready to return to this comment thread and provide an update that it still wasn't working, but as I was retracing my steps while writing up the comment, I noticed the words "in this environment" in the description of "access_token".  

That's when I realized that I was working in the "No environment" environment.  I guessed that maybe it needed a specific environment, so I created a new environment, added the variable, went to the request A, cleared and refetched the token, then went to request B, clicked on 'access_token' and I was thrilled to see the Live Preview started showing the token!

So my suggestions are:

1. If possible, let the user choose a specific request from which to get the token, in the form of a drop-down on the 'access_token' settings, similar to Insomnia's "Request -> Body Attribute" settings.  This would provide two benefits: (1) if the user has multiple requests all with different OAuth credentials, requesting one OAuth token won't interfere with other requests that might use a different token from a prior request, and (2) would essentially eliminate the need for an environment variable.

2. If #1 is not possible, or if the plugin still wants to support "last token" access via an environment variable, then if possible, support the 'No Environment' environment.  If supporting the "No Environment" isn't possible, then definitely call that out in the documentation.

So hopefully that wasn't too rambling!

**Comment by kitingChris** [comment]
Not sure if I understand your point right but when building an api do you really need more than one identity provider?
Or why is your need for multiple requests all with different OAuth credentials?
To support multiple credential providers it not only would need a major rewrite of this plugin but also some kind of dropdown functionality which is as far as I know not offered by any of insomnias plugin api...

The 'No Environment' enviroment might work. But only if you do not have to modify the 'oauth2_token_url_suffix' since the only possibility to have some kind of configuration is via the environment... Or at least you could try to configure the variable as Base Environment... ?


But still unsure what your want to achieve and if it might exceed the prupose of the plugin.

It is intended to have one pseudo login request in which you can get an access_token from the oauth handle and reuse it in other requests. 

Actually the plugin itself is not a official product of our company but only a sideproduct since we are currently in one of our projects migrating from postman to insomnia and missing the oauth feature. 

But when I find some time I will enhance the Readme! :)

**Comment by flmbray** [comment]
I mostly use Insomnia for testing APIs that our application uses, not so much for building an API, although our application includes an API that I test as well.  The APIs that I test are sometimes completely separate with no relation among them, but not always.  For example, our application basically runs our internal quoting and purchasing tool, among 1000 other things.  An account manager might first use a web-based tool provided by a manufacturer to create a quote for a product suite (think of something like buying a computer from a manufacturer, where you have to choose a lot of different parts - keyboard, mouse, monitor, etc, but where the parts could in fact be purchased from many different suppliers.)  We first access the manufacturer's API to download the quote into our internal tool, then we access the API for our internal ERP system to see if we have details on each of the parts from various suppliers, and depending on the response we might then query missing suppliers to see if they have the part, or suppliers that we know carry the part to find out what quantity they have in stock and at what price, etc.  Each link in that process requires separate authentication credentials, which is compounded even more when accessing each of the different suppliers.  I use the data linking feature (is that the right name?) so that I can test this process in Insomnia chained end-to-end without having to cut/paste between different requests.

Admittedly the OAuth separation needed by your plugin could be achieved by creating separate environments for each connection (and the data-linking would still work), but I already use Insomnia environments to separate our internal environments - that is, there are environments for development, QA testing, user testing, demo, production simulation, and production.  Creating a separate Insomnia environment for each API that I need to access with OAuth inside of each of these internal environments would be very cumbersome to configure and operate.  

However...  Keep in mind that I was researching this issue for another user, with different needs.  I am totally happy having found a plugin that was even close to what she needed.  The example I gave above is actually somewhat contrived because I'm not totally sure what her actual use case was - I was only following her stated requirements, perhaps too literally.  I myself don't actually use OAuth authentication in the way I described, where the OAuth is configured on the Authentication tab.  Instead, I execute the authentication as a completely separate request, and then use the "Body Response attribute" linking to get the token into the header of other requests.  In fact, I knew that Insomnia already provides a "Trigger Behavior" to trigger a linked request, but it seems there is an option to refresh "When Expired" that I don't recall being there originally.  I haven't tested it much yet, but I see two potential problems with it - first, you have to manually configure the timeout - it would be better if the timeout could be pulled from the request itself - and second, it probably doesn't honor the current environment, so it will probably use a "test" token in a "production" environment were I to execute such a request before the expiration.  

I completely understand if the current plugin capabilities meet your own needs and none of the above is sufficient reasoning for you to change it, or even if it is but you don't have time or desire to change it.  I certainly had no expectation that you would - my initial request was a request for help to get it to work, for my other user.  But if any of the above helps improve your plugin (maybe you could address the two shortcomings of the other method above) then I'm happy to have been of service!

**Comment by glconti** [comment]
Hi, I wanted just to add here my personal story, also for me the plugin didn't seem to work in any way, but in the end, it was only due to my little knowledge of the Insomnia wording/features. Maybe is this the simplified version of what happened to @flmbray ?

From the readme:
> Use the following custom template tag: `access_token`

And in this small sentence there's the whole difference, this plugin installs in fact a **template tag** which is something really different form an **environment variable** which I was mistakenly trying to use!

In fact I was trying to use an env variable in my request (note the purple color):

> {{ access_Token}}

![image](https://user-images.githubusercontent.com/13535297/196718709-e3df9f3d-f263-4f7e-a4c7-a90740ff0f2d.png)

But, I learned the hard way, I had to use a **template tag**, invoking the list of **template tags** with `ctrl + space` (note the blue color):

![image](https://user-images.githubusercontent.com/13535297/196719222-757ccd81-a2ab-4bb5-b88c-84b7cc8d4e68.png)
![image](https://user-images.githubusercontent.com/13535297/196719265-2bccaa90-c24e-401c-a7d7-158804f9f86a.png)

And it is true that you really don't need to have any environment at all.

Maybe these instructions could be added to the readme for all other newcomers to Insomnia.
